lib/CL/pocl_llvm_api.cc

```diff
@@ -1527,10 +1561,37 @@ static PassManager& kernel_compiler_passes
   passes.push_back("domtree");
   if (device->autolocals_to_args)
 	  passes.push_back("automatic-locals");
+
   if (SPMDDevice)
     passes.push_back("flatten-inline-all");
-  else
-    passes.push_back("flatten-globals");
+  else {
+    /* LLVM is either still not intelligent enough to flatten,
+     * or we're missing some pass.
+     *
+     * The only reason we've reintroduced "global-only" flatten was that
```

Comment:
> only those that have to be (functions which take in the context struct basically)

That's what i meant by "global-only".

> Could we simply rely on LLVM inliner nowadays?

The performance will take a hit (in some cases significant, see the entire comment), so i'd prefer to play around with LLVM inliner **after** we do the release. The way it's set up with this commit, it's at least as good as pocl 0.14 and it also works with read/write_image{f,i,ui}.